src/tools/tidy/src/error_codes_check.rs

```diff
@@ -0,0 +1,183 @@
+                    error_codes.insert(err_code, 0);
+                }
+            }
+        } else if s.starts_with("```") && s.contains("compile_fail") && s.contains('E') {
+            if let Some(err_code) = s.splitn(2, 'E').skip(1).next() {
+                if let Some(err_code) = err_code.splitn(2, ',').next() {
+                    let nb = error_codes.entry(format!("E{}", err_code)).or_insert(0);
+                    *nb += 1;
+                }
+            }
+        } else if s == ";" {
+            reached_no_explanation = true;
+        } else if reached_no_explanation && s.starts_with('E') {
+            if let Some(err_code) = s.splitn(2, ',').next() {
+                let err_code = err_code.to_owned();
+                if !error_codes.contains_key(&err_code) { // this check should *never* fail!
+                    error_codes.insert(err_code, 0);
+                }
+            }
+        } else if s.starts_with("#### Note: this error code is no longer emitted by the compiler") {
+            if let Some(last) = last_error_code {
+                error_codes.get_mut(&last).map(|x| *x += 1);
+            }
+            last_error_code = None;
+        }
+        buf.clear();
+    }
+}
+
+fn extract_error_codes_from_tests(
+    mut f: BufReader<File>,
+    mut buf: String,
+    error_codes: &mut HashMap<String, usize>,
+) {
+    while let Ok(bytes_read) = f.read_line(&mut buf) {
+        if bytes_read == 0 {
+            return;
+        }
+        let s = buf.trim();
+        if s.starts_with("error[E") || s.starts_with("warning[E") {
+            if let Some(err_code) = s.splitn(2, ']').next() {
+                if let Some(err_code) = err_code.splitn(2, '[').skip(1).next() {
+                    let nb = error_codes.entry(err_code.to_owned()).or_insert(0);
+                    *nb += 1;
+                }
+            }
+        }
+        buf.clear();
+    }
+}
+
+pub fn check(path: &Path, bad: &mut bool) {
+    println!("Checking which error codes lack tests...");
+    let mut error_codes: HashMap<String, usize> = HashMap::new();
+    let _ = visit_dirs(path, 0, &mut |entry: &DirEntry| {
+        if let Ok(file_name) = entry.file_name().into_string() {
+            if file_name == "error_codes.rs" {
+                open_file(&entry.path(), &mut |f, buf| {
+                    extract_error_codes(f, buf, &mut error_codes);
+                });
+            } else if file_name.ends_with(".stderr") {
+                open_file(&entry.path(), &mut |f, buf| {
+                    extract_error_codes_from_tests(f, buf, &mut error_codes);
+                });
+            }
+        }
+    });
+    println!("Found {} error codes", error_codes.len());
+
+    let mut errors = Vec::new();
+    for (err_code, nb) in &error_codes {
+        if *nb == 0 && !WHITELIST.contains(&err_code.as_str()) {
+            errors.push(format!("Error code {} needs to have at least one UI test!", err_code));
+        }
+    }
+    errors.sort();
```

Comment:
This isn't really all that helpful I imagine to get these in a sorted order -- most of the time, you're probably only going to see one or two -- can we just print immediately? If we care so much, then we can have `error_codes` be a `BTreeSet` which we delete items from when finding the error code in tests.

Reply:
No it's **very** useful since they are not sorted at all. For example, if you read an `error_codes.rs` file and it has both long error explanations and error code declaration, the order is already broken. Also, please keep in mind that we also read from two different contexts: ui tests and long error explanations. If we remove it because we found a test and then find it again with no test (just an error code declaration for instance), then it'll throw an error.

Reply:
It also made my life a lot simpler when I had to go through spotted errors. :)

Reply:
I don't quite understand why sorting is useful (since we're sorting by error code, which has no relation to where the code is declared, at least for the most part); however, if you feel it is, then that's fine. I would still like to see us move to the BTreeSet and removal from it as we find tests, since I think that'll be (much) cleaner.

Reply:
We can't move to a BTreeSet. For example:

1. You find the ui test file with the error code, no need to add it into the BTreeSet so we continue.
2. You find the same error code declaration (so no long error explanation and therefore, no test) in an `error_codes.rs` file. No test, you add it into the BTreeSet.
3. The check fails because this error code (which has a test) is detected as not having tests.

Or maybe I misunderstood what you tried to explain to me?

Reply:
Ah, I was expecting to do this in two passes - discover error codes in one pass and then use that information. But since we're not already doing that lets leave it as is for now.